### PR TITLE
Migrate code from ossf/omega-triage-portal to alpha-omega.

### PR DESCRIPTION
**Pull request by scovetta**
This PR migrates code from the private ossf/omega-triage-portal repository to alpha-omega.

The project is not intended for public use, and is very much "in early development".

Signed-off-by: Michael Scovetta <michael.scovetta@gmail.com>

### CODE REVIEW

**Comment by scovetta** [comment]
Basic view of the home page:
image

File: omega/triage-portal/src/.env-template [hunk 1]
```diff
@@ -0,0 +1,27 @@
+BASE_URL='https://localhost'
+
+# Re-generate using `python -c "import secrets; print(secrets.token_hex(64))"`
+SECRET_KEY = "%RANDOM_STRING%"
+
+DEBUG=True
+
+# Database
+DATABASE_ENGINE='django.db.backends.postgresql_psycopg2'
+DATABASE_NAME='triage'
```
Comment (Cyber-JiuJiteria) [inline omega/triage-portal/src/.env-template:10]: Values seem to be from local dev env. Future enhancements will need to ensure sensitive credentials are not shared.

File: omega/triage-portal/src/core/__init__.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        if optional:
+            return False
+        raise ImproperlyConfigured(
+            f"Error: You must set the {var_name} environment variable."
+        ) from ex
+
+
+def to_bool(option: str) -> bool:
+    """Convert a string to a boolean."""
+    return option and option.lower().strip() in ["true", "1"]
```
Comment (Cyber-JiuJiteria) [inline omega/triage-portal/src/core/__init__.py:24]: curious of `false` or `0` use case

File: omega/triage-portal/src/triage/management/commands/import_toolshed.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        scan_map = {}
+
+        if package:
+            prefix = Command.prefix_for_package_url(package)
+            blobs = self.container.list_blobs(name_starts_with=prefix)
+        elif options.get("import_all"):
+            package_url = None
+            blobs = self.container.list_blobs(
+                name_starts_with="npm",
+            )  # TODO: temporary
```
Comment (Cyber-JiuJiteria) [inline omega/triage-portal/src/triage/management/commands/import_toolshed.py:72]: What is the thought behind the #TODO temporary?

File: omega/triage-portal/src/triage/models/case.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    """
+
+    class CasePartner(models.TextChoices):
+        NONE = "N", _("None")
+        GITHUB_SECURITY_LAB = "GS", _("GitHub Security Lab")
+        CERT = "CT", _("CERT")
+        MSRC = "MS", _("Microsoft Security Response Center")
+        NOT_SPECIFIED = "NS", _("Not Specified")
+
+    uuid = models.UUIDField(default=uuid.uuid4, editable=False, db_index=True)
```
Comment (Cyber-JiuJiteria) [inline omega/triage-portal/src/triage/models/case.py:25]: similar to previous comment. Should we be using the unique=True to ensure there are no duplicates

File: omega/triage-portal/src/triage/models/tool_defect.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            .get_queryset()
+            .filter(
+                state__in=[WorkItemState.NEW, WorkItemState.ACTIVE, WorkItemState.NOT_SPECIFIED]
+            )
+        )
+
+
+class ToolDefect(BaseTimestampedModel, BaseUserTrackedModel):
+    """
+    A tool defect is a defect that is filed against a tool.
```
Comment (Cyber-JiuJiteria) [inline omega/triage-portal/src/triage/models/tool_defect.py:34]: Are these defect found by the Omega analyzer, scan result, vulnerabilties/CVEs, or any type?

File: omega/triage-portal/src/triage/models/triage.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+class TriageRule(models.Model):
+    """
+    ???
+    def applies(finding) -> bool
+    def action(finding) -> void
+
+    if applies(f): action(f)
+    """
+
+    class TriageEvent(models.TextChoices):
```
Comment (Cyber-JiuJiteria) [inline omega/triage-portal/src/triage/models/triage.py:20]: By TriageEvent, is this the triage workflow events, like research, documenting, in progress, under review?

Or the state of a TriageEvent as far as updates and auditing changes?

File: omega/triage-portal/src/triage/models/wiki.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+from django.db import models
+
+from triage.models import BaseTimestampedModel, BaseUserTrackedModel, WorkItemState
+
+logger = logging.getLogger(__name__)
+
+
+class WikiArticleRevision(BaseTimestampedModel, BaseUserTrackedModel):
+    uuid = models.UUIDField(default=uuid.uuid4, editable=False, db_index=True)
+    article = models.ForeignKey("WikiArticle", on_delete=models.CASCADE, related_name="revisions")
```
Comment (Cyber-JiuJiteria) [inline omega/triage-portal/src/triage/models/wiki.py:13]: What are the articles? 

markdown files, pdf uploads, links to URLs, or textbox for writing?

File: omega/triage-portal/src/triage/static/triage/omega.js [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+                    row: file_line - 1,
+                    column: 0,
+                    text: finding_title,
+                    type: 'error'
+                }]);
+            } else {
+                ace.edit('editor').getSession().clearAnnotations();
+            }
+
+            // @TODO Are these necessary?
```
Comment (Cyber-JiuJiteria) [inline omega/triage-portal/src/triage/static/triage/omega.js:133]: Todo left behind. Are they necessary?

File: omega/triage-portal/src/triage/util/azure_blob_storage.py [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+        if package_url.namespace:
+            parts = [package_url.type, package_url.namespace, package_url.name, package_url.version]
+        else:
+            parts = [package_url.type, package_url.name, package_url.version]
+
+        # Escape reserved URL characters
+        prefix = "/".join(map(urllib.parse.quote, parts))
+
+        # TODO: Add validation based on https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata
```
Comment (Cyber-JiuJiteria) [inline omega/triage-portal/src/triage/util/azure_blob_storage.py:114]: todo for validation on blobs

File: omega/triage-portal/src/triage/util/azure_blob_storage.py [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        except:
+            logger.exception("Failed to get blob contents")
+            cache.set(cache_key, "", timeout=DEFAULT_CACHE_TIMEOUT)
+            return None
+
+    def get_all_files(self):
+        """Retrieve all files about the scan."""
+        return self.get_tool_files() + self.get_package_files() + self.get_intermediate_files()
+
+    def get_intermediate_files(self):
```
Comment (Cyber-JiuJiteria) [inline omega/triage-portal/src/triage/util/azure_blob_storage.py:210]: get_intermediate_files() to be implemented. Issue Created

File: omega/triage-portal/src/triage/util/finding_importers/sarif_importer.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+                    # Ensure we only insert the same message once
+                    key = {
+                        "title": message,
+                        "path": uri,
+                        "line_number": get_complex(location, "physicalLocation.region.startLine"),
+                    }
+                    key = hashlib.sha256(json.dumps(key).encode("utf-8")).digest()
+
+                    if key not in processed:
```
Comment (Cyber-JiuJiteria) [inline omega/triage-portal/src/triage/util/finding_importers/sarif_importer.py:112]: Duplicates are checked within the set of the finding imported, however, should we also considered finding already in the database?

File: omega/triage-portal/src/triage/util/search_parser.py [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+        pp.Keyword("updated").suppress()
+        + pp.Literal(":").suppress()
+        + pp.one_of(["<", ">", "<=", ">=", "=="]).setResultsName("op")
+        + pp.pyparsing_common.iso8601_date("datetime")
+    ).setResultsName("updated_dt")
+
+    created_dt_clause = pp.Group(
+        pp.Keyword("created").suppress()
+        + pp.Literal(":").suppress()
+        + pp.one_of(["<", ">", "<=", ">=", "=="]).setResultsName("op")
```
Comment (Cyber-JiuJiteria) [inline omega/triage-portal/src/triage/util/search_parser.py:69]: Would there be a case of `!=`?

File: omega/triage-portal/src/triage/util/source_viewer/__init__.py [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+    """
+    if not files:
+        return []
+
+    result = []
+    seen_nids = set()
+    if root:
+        result.append(
+            {
+                "id": root,  # TODO minimize this via a lookup table
```
Comment (Cyber-JiuJiteria) [inline omega/triage-portal/src/triage/util/source_viewer/__init__.py:41]: can you elaborate on this TODO so an issue can be created?

File: omega/triage-portal/src/triage/views/home.py [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+    most_recent_case = Case.objects.all().order_by("-updated_at")
+    case_last_updated = most_recent_case.first().created_at if most_recent_case else None
+
+    my_work = {
+        "num_cases": Case.objects.filter(assigned_to=request.user).count(),
+        "num_findings": Finding.objects.filter(assigned_to=request.user).count(),
+        "num_tool_defects": ToolDefect.objects.filter(assigned_to=request.user).count(),
+    }
+    metrics = {
```
Comment (Cyber-JiuJiteria) [inline omega/triage-portal/src/triage/views/home.py:24]: additional metrics:
- resolved findings?
- updated findings? (say a finding was patched, but showed up again in a later report)

**Review by Cyber-JiuJiteria — COMMENTED** [review]
Completed review of all the new files. Added comments and created additional issues to reflect design suggestions, not yet implemented, and left behind todos. Will approve once reviewed.

File: omega/triage-portal/src/triage/models/case.py (hunk 4 above)
Comment (scovetta) [reply omega/triage-portal/src/triage/models/case.py]: Good catch, fixed (along with all of the other UUIDFields).

Comment (Cyber-JiuJiteria) [reply omega/triage-portal/src/triage/models/case.py]: ack resolving

File: omega/triage-portal/src/triage/models/tool_defect.py (hunk 5 above)
Comment (Cyber-JiuJiteria) [reply omega/triage-portal/src/triage/models/tool_defect.py:34]: so anything we define to be a tool, within the T-P and Analyzer, is what we are considering to be ToolDefects?

File: omega/triage-portal/src/triage/models/wiki.py (hunk 7 above)
Comment (Cyber-JiuJiteria) [reply omega/triage-portal/src/triage/models/wiki.py:13]: Should we have a limit on characters for the text? Is this similar to notes or comments? Or more as the "final report"

File: omega/triage-portal/src/triage/util/search_parser.py (hunk 12 above)
Comment (scovetta) [reply omega/triage-portal/src/triage/util/search_parser.py]: Good catch, added `!=` to created_dt, updated_dt, and priority.

File: omega/triage-portal/src/triage/views/home.py (hunk 14 above)
Comment (scovetta) [reply omega/triage-portal/src/triage/views/home.py:24]: Good ideas, created #84 to track.

File: omega/triage-portal/src/triage/util/source_viewer/__init__.py (hunk 13 above)
Comment (scovetta) [reply omega/triage-portal/src/triage/util/source_viewer/__init__.py:41]: I should have added more information when I wrote it. Past-me certainly knew why, today-me has no clue.

File: omega/triage-portal/src/triage/util/finding_importers/sarif_importer.py (hunk 11 above)
Comment (scovetta) [reply omega/triage-portal/src/triage/util/finding_importers/sarif_importer.py:112]: Yep, we check the database later on around line 147:

```python
                        if Finding.objects.filter(
                            title=finding.title,
                            file=finding.file,
                            file_line=finding.file_line,
                            project_version=finding.project_version,
                        ).exists():
                            logger.debug("Duplicate finding, skipping.")
                            continue
```

File: omega/triage-portal/src/triage/templates/triage/filter_show.html [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+                </div>
+            </div>
+        </div>
+    </form>
+</div>
+
+    {% endblock %}
+{% endblock %}
+
+{% block javascript %}
```
[the remark this answers is not in the document]
Comment (scovetta) [reply omega/triage-portal/src/triage/templates/triage/filter_show.html:89]: Yep, we probably should -- though it's sometimes nice to keep page-specific logic next to the page itself. We can optimize away later by packing/compressing/etc., but if you'd prefer to keep the html-ish content and JS content separate, that's totally fine.

File: omega/triage-portal/src/triage/static/triage/omega.js (hunk 8 above)
Comment (scovetta) [reply omega/triage-portal/src/triage/static/triage/omega.js]: Removed -- couldn't find any references to what they were referring to, so probably refactored away at some point in the past.

File: omega/triage-portal/src/triage/models/tool_defect.py (hunk 5 above)
Comment (scovetta) [reply omega/triage-portal/src/triage/models/tool_defect.py:34]: Correct -- all ToolDefects are mapped to a Tool, and a Tool would be something like CodeQL (or CodeQL version X). There are different types of Tools though:
image

So I think we can cover pretty much any "finding generated by <X>" thing.

File: omega/triage-portal/src/triage/management/commands/import_toolshed.py (hunk 3 above)
Comment (scovetta) [reply omega/triage-portal/src/triage/management/commands/import_toolshed.py:72]: This code exists to import (en masse) results generated by the Analyzer, stored in a blob store. But any type of package can be used (not just npm). I think I have npm in there for testing -- this is one of the areas that needs a bit of work.

File: omega/triage-portal/src/core/__init__.py (hunk 2 above)
Comment (scovetta) [reply omega/triage-portal/src/core/__init__.py:24]: I went with only True, true, TRUE, 1, "1", etc. convert to boolean 'true', and everything else (either 0/false or "foo") all convert to boolean 'false'.

It's more relaxed than https://docs.python.org/3/distutils/apiref.html#distutils.util.strtobool which raises ValueError on anything like "foo". We can change it, but this function is only used in settings to read environment variables and booleanize them.

File: omega/triage-portal/src/.env-template (hunk 1 above)
Comment (scovetta) [reply omega/triage-portal/src/.env-template:10]: Yep, good catch!

File: omega/triage-portal/src/core/__init__.py (hunk 2 above)
Comment (Cyber-JiuJiteria) [reply omega/triage-portal/src/core/__init__.py:24]: No need to change, hadnt see this before and wanted to ensure all use cases were covered.

File: omega/triage-portal/src/triage/management/commands/import_toolshed.py (hunk 3 above)
Comment (Cyber-JiuJiteria) [reply omega/triage-portal/src/triage/management/commands/import_toolshed.py:72]: Gotcha. we'd have to test other package types and ensure functionality on the import. Assuming only import on this area of the code.

File: omega/triage-portal/src/triage/models/tool_defect.py (hunk 5 above)
Comment (Cyber-JiuJiteria) [reply omega/triage-portal/src/triage/models/tool_defect.py:34]: ack. i like this. thank you for clarification. resolving.

File: omega/triage-portal/src/triage/templates/triage/filter_show.html (hunk 15 above)
[the remark this answers is not in the document]
Comment (Cyber-JiuJiteria) [reply omega/triage-portal/src/triage/templates/triage/filter_show.html:89]: Could be on the lower end of priorities, but future designs should separate, especially with the consideration that triage-portal will be publicly accessible on the web. We'd need to evaluate risk for that considering the sensitivity of the data in T-P

File: omega/triage-portal/src/triage/util/finding_importers/sarif_importer.py (hunk 11 above)
Comment (Cyber-JiuJiteria) [reply omega/triage-portal/src/triage/util/finding_importers/sarif_importer.py:108]: nice! resolving. thank you

File: omega/triage-portal/src/.env-template (hunk 1 above)
Comment (Cyber-JiuJiteria) [reply omega/triage-portal/src/.env-template:10]: ack. resolving.

File: omega/triage-portal/src/triage/util/source_viewer/__init__.py (hunk 13 above)
Comment (Cyber-JiuJiteria) [reply omega/triage-portal/src/triage/util/source_viewer/__init__.py:41]: Might need a time machine. Should we create an issue or leave as is?

**Comment by Cyber-JiuJiteria** [comment]
a few more comments left for review before approving

File: omega/triage-portal/src/triage/models/wiki.py (hunk 7 above)
Comment (scovetta) [reply omega/triage-portal/src/triage/models/wiki.py:13]: We could always add a limit later, but only trusted folks will be adding content, so maybe "wiki" was the wrong word in the first place.

File: omega/triage-portal/src/triage/util/source_viewer/__init__.py (hunk 13 above)
Comment (scovetta) [reply omega/triage-portal/src/triage/util/source_viewer/__init__.py]: Let's just leave as is -- I cleaned up the comment.

File: omega/triage-portal/src/triage/models/triage.py (hunk 6 above)
Comment (scovetta) [reply omega/triage-portal/src/triage/models/triage.py:20]: A TriageEvent is intended to occur when something interesting happens, like when a new finding is added, or when a finding's triage status is changed. But we haven't hooked this up to anything, so it doesn't actually do anything yet.

**Comment by scovetta** [comment]
@Cyber-JiuJiteria I think I got to all of your comments. :-)

File: omega/triage-portal/src/triage/models/triage.py (hunk 6 above)
Comment (Cyber-JiuJiteria) [reply omega/triage-portal/src/triage/models/triage.py:20]: ack and adding to milestone doc. resolving.

File: omega/triage-portal/src/triage/models/wiki.py (hunk 7 above)
Comment (Cyber-JiuJiteria) [reply omega/triage-portal/src/triage/models/wiki.py:13]: so the intention was more of adding some sort of content, rather than a "wiki". Resolving.

**Review by Cyber-JiuJiteria — APPROVED** [review]

**Comment by Cyber-JiuJiteria** [comment]
Approved and merging.

**Pull request merged**